Post about uncertainties

mmattamala opened the pull request:
Hi @dellaert 

I wrote this post based on the discussions we had in the gtsam-user groups about the uncertainties. I tried to cover almost all the topics we discussed, and creating figures to support the sections.

Let me know your comments, I'll be happy to address any concern. I will also ask the people here at DRS to get more feedback.

Additionally, there was one technical issue with Latex integration I couldn't solve: using the `\cancel{}` command. I checked everything locally to see if things looked good, but I couldn't make that one work.

Thanks!

dellaert reviewed (commented):
Wow @mmattamala ! This looks amazing. I think this could be published as is, although I would some time to fully read. This leads me to two main comments/suggestions that would help me:

1) it's long :-) Perhaps this could be split into a part with linear/nonlinear, and a part with manifolds.
2) more controversial/involved: in GTSAM I deliberately use |h(x)-z|, not |z-h(x)|. Then linearization is |x0+H dx-z| = |H dx - (z-x0)|. I noticed you are going the other way :-)

I have not read the post in detail yet, it's something I noticed and wanted to start a conversation. You worked on this for a long time, and I'm on board to publish as is (after I read in detail). But the above might help people grok everything.

mmattamala commented on _posts/2021-02-07-uncertainties.md:
```diff
@@ -0,0 +1,766 @@
+\end{equation}
+$$
+
+However, we now that in reality things do not work that way, and we will usually have errors produced by noise in our sensors. The most common way to address this problem is adding some *zero-mean Gaussian noise* $$\eta_i\sim Gaussian(\mathbf{0}_{2\times1}, \Sigma_i)$$ to our measurement to model this uncertainty:
```
Just fixed, thanks for noticing this :)

mmattamala commented:
Hi @dellaert , glad to know that you liked it!

Regarding the comments:
1. I'm happy to split it if it works better. I guess we would need a summary at the end to the first part to connect to the next.
2. I can change the residual as you mentioned to be consistent, it's an easy fix but it must be done carefully.

I'm looking forward to any other comments or suggestions, I really hope the post helps :)

jlblancoc commented:
Amazing work, Matias!

I'm also more used to see h(x)-z, which also doesn't flip the sign of the
observation function Jacobian. Just a convention, but I think it's slightly
more convenient.... plus, it's the one used in the library! :-)

dellaert commented:
cool, re-request a review then I’ll get pinged.

mmattamala commented:
Thanks a lot @jlblancoc and @dellaert 

I'd like to ask you a question I thought it could be relevant to add in the "linear and nonlinear" part:

**Do you have any idea on how the covariance estimate gets affected when we use Levenberg-Marquardt?** Because the damping is changing the Hessian, hence the covariance estimate should be different.

I know that apart from the convergence behaviour, LM can help when we have some ill-conditions in the problem, and that sometimes work as an "easy fix". But I'm afraid that when we want to extract the covariance as well, it will get over/underestimated (I have not tested it but perhaps you have :) ). 

It may be useful to mention that "issue", and that instead of just changing the optimizer to solve ill-conditions, it's better to check the jacobians or adding explicit prior factors to improve the system but still having some control on our uncertainty about the variables. What do you think?

dellaert commented:
LM just finds a minimum. Covariance calculations are based on non-damped Jacobians. But, you should never use LM to “abide” an ill-posed problem in the first place :-)

michaelbosse commented on _posts/2021-02-07-uncertainties-part2.md:
```diff
@@ -0,0 +1,631 @@
+
+In these posts we took a long trip to explain how the covariances are introduced and extracted from the estimation framework. We reviewed how a clear understanding of the conventions (explicit or not) embeded in our factor graph problems is fundamental to make sense of the quantities involved. We introduced the concept of _right-hand_ and _left-hand_ conventions which, while not standard, allowed ...
+
+We also showed how the right-hand convention is related to how the composition is defined when working with manifolds, such as rigid-body transformations and rotation matrices. The  **local** and **retract** operations are also defined from the right-hand side, indicating that the increments and the covariances we compute are defined with respect to the base frame.
```
should be "... when working with *groups*"

It would be good to the concepts for manipulating transformations, ie generating odometry factors, distinct from representing uncertainty on these group objects.   For poses in gtsam both are very similar and use a right handed convention, but understanding their distinction can help with applying these ideas and understanding a variety of nonlinear objects used in gtsam such  as unit vectors, nav states (pose + velocity),  camera calibration matrices, etc.

michaelbosse reviewed (changes requested):
In general I believe this is a very good and detailed write up.  I just have some comments that address distinctions between groups and manifolds.

mmattamala commented:
Thanks a lot for the comments @michaelbosse 

I read them carefully and will update the post to be more clear about the distinctions you mentioned. I think the trickiest part is to keep the differences clear so as to understand when each concept applies (uncertainties for manifolds, frame conventions for groups, adjoints for Lie groups)

dellaert commented:
Thanks Mike!!
Matias, the following document, which is more implementation oriented, might also help, maybe you could link to it: https://gtsam.org/notes/GTSAM-Concepts.html

mmattamala commented:
Hi @dellaert 

Thanks for the previous explanation on LM and the "Concepts" post.
I updated the post after Mike's feedback, as well as José Luis' and your comments. It is a 3 parts post now, I hope it is not a problem :)

- **Part 1, linear and nonlinear:** how we get the covariances from the Hessian
- **Part 2, frames and manifolds:** introduces the reference frames and the operations on groups and manifolds following the frames and the "concepts" post.
- **Part 3, adjoints and covariances:** the adjoint as a way to relate increments on the left and the right (including what Mike mentioned before in his first 2 comments), and Mangelson's formulas compliant with GTSAM.

dellaert reviewed (approved):
Looks amazing to me. Still some small comments, and I attach he rendered pdfs so Mike can approve if he is ok with your changes:
1) in part 1 the "- noise" still feels awkward, and I don't see actually where you need it. You could just say "Ax-b ~ normal"....
2) in part 2 (which is beautiful!) formula 11 renders weirdly for me, with a red "\cancel" in subscript. See attached pdf.

[Reducing the uncertainty about the uncertainties, part 1_ Linear and nonlinear _ GTSAM.pdf](https://github.com/borglab/gtsam.org/files/5984605/Reducing.the.uncertainty.about.the.uncertainties.part.1_.Linear.and.nonlinear._.GTSAM.pdf)

[Reducing the uncertainty about the uncertainties, part 2_ Frames and manifolds _ GTSAM.pdf](https://github.com/borglab/gtsam.org/files/5984604/Reducing.the.uncertainty.about.the.uncertainties.part.2_.Frames.and.manifolds._.GTSAM.pdf)

[Reducing the uncertainty about the uncertainties, part 3_ Adjoints and covariances _ GTSAM.pdf](https://github.com/borglab/gtsam.org/files/5984595/Reducing.the.uncertainty.about.the.uncertainties.part.3_.Adjoints.and.covariances._.GTSAM.pdf)

johnwlambert commented on _posts/2021-02-07-uncertainties-part1.md:
```diff
@@ -0,0 +1,203 @@
+
+where the vector $\mathcal{X}^{*}$ stacks all the variables in the problem, $\mathbf{A}$ is a matrix that stacks the linear terms in the factors, and $\mathbf{b}$ does the same for the constant vector terms.
+
+First point to note here is that finding the solution $\mathcal{X}^{*}$ requires to invert the matrix $\mathbf{A}^{T} \Sigma^{-1} \mathbf{A}$, which in general is hard since it can be huge and dense in some parts. However we know that there are clever ways to solve it, such as iSAM and iSAM2 that GTSAM already implements, which are covered in [this comprehensive article](https://www.cc.gatech.edu ...
```
Very nice post @mmattamala. One small typo here in spelling of the authors' names

mmattamala replied:
Thanks John for the comment and proofreading!

johnwlambert commented on _posts/2021-02-07-uncertainties-part1.md:
```diff
@@ -0,0 +1,203 @@
+
+In this first post we reviewed the basics ideas to solve factor graphs and extract estimates from them. We showed how the linear problem described by the *normal equations* not only allows us to find the solution, but also encodes information about its covariance - given by inverse of the Fisher information matrix. Additionally, in the context of nonlinear problems, we showed how **such covarianc ...
+
+In our [next post](https://gtsam.org/2021/02/07/uncertainties-part2.html) we will review how these ideas generalize when we are not estimating *vector* variables anymore but other objects such as rotation matrices and rigid-body transformations - generally known as **manifolds**. While this will introduce a bit of complexity no only in the maths but also the notation, by being explicit about our  ...
```
perhaps a small typo here: "no only in the maths" -> "not only in the maths"

johnwlambert commented on _posts/2021-02-07-uncertainties-part2.md:
```diff
@@ -0,0 +1,500 @@
+
+Here we made explicit all the frames involved in the transformations we usually can find in textbooks: An homogeneous point $${_{W}}\mathbf{P}$$ expressed in the world frame $W$, is transformed by the *extrinsic calibration matrix* $$\mathbf{T}_{CW}$$ (a rigid body transformation ) which represents the world $W$ in the camera frame $C$, producing a vector $${_{C}}\mathbf{P}$$ in the camera frame  ...
+
+Since in GTSAM the extrinsic calibration is defined the other way to be consitent with the right-hand convention, i.e $\mathbf{T}_{WC}$, the implementation of the `CalibratedCamera` handles it by [properly inverting the matrix before doing the projection](https://github.com/borglab/gtsam/blob/develop/gtsam/geometry/CalibratedCamera.cpp#L120) as we would expect:
```
Another nice post! Small typo here "consitent" -> "consistent"

mmattamala commented:
Thanks a lot @dellaert and @johnwlambert for your feedback!

Frank, regarding your comments:
1. Yeah, I see your point with the "negative noise". I just wanted to make the manipulation explicit since it's the same we do later but it looks odd indeed. Another approach would be to isolate it such as keep it positive, but when we define the factor we can swap the model and the measurement as usual to keep your convention (because it's more convenient).
2. In part 2, now I figured out how to solve the problem by adding a macro in the MathJax configuration.

michaelbosse commented on _posts/2021-02-07-uncertainties-part2.md:
```diff
@@ -0,0 +1,498 @@
+
+As we briefly mentioned before, objects such as rotation matrices are difficult to manipulate in the estimation framework because they are matrices. A 3D rotation matrix $\mathbf{R}$ represents 3 orientations with respect to a reference frame but, in raw terms, they are using 9 values to do so, which seems to *overparametrize* the object. However, the constraints that define a rotation matrix -an ...
+
+**That is what makes working with manifolds so convenient**: All the constraints that are part of the definition of the object are naturally handled, and we can work in tangent vector spaces using their *inherent* dimension. The same happens for rigid-body transformations (6 dimensions represented by a 16 elements matrix), quaternions (3 orientations represented by a 4D vector), **and even object ...
```
Calibration matrices *are* in fact groups ;-)   See https://en.wikipedia.org/wiki/Triangular_matrix

Perhaps use the example of Unit Vectors as a non-group object (which is still a manifold)

mmattamala replied:
Great, thanks for clarifying this and the example with the unit vectors :)

michaelbosse commented on _posts/2021-02-07-uncertainties-part2.md:
```diff
@@ -0,0 +1,498 @@
+
+In order to work with manifolds, we need to define 2 operations, which are the key to transform objects between them and the tangent spaces:
+
+1. **Local**: An operation that maps elements from the manifold to the tangent space $\mathbf{\xi} = \text{local}(\mathbf{T})$
```
This single parameter version of **local** assumes the existence of an Identity object, which is only true for groups.
The more general local operation takes the difference between two objects and expresses it as a vector in the tangent space of one

mmattamala replied:
I changed the notation a bit to make this more explicit.

michaelbosse commented on _posts/2021-02-07-uncertainties-part2.md:
```diff
@@ -0,0 +1,498 @@
+In order to work with manifolds, we need to define 2 operations, which are the key to transform objects between them and the tangent spaces:
+
+1. **Local**: An operation that maps elements from the manifold to the tangent space $\mathbf{\xi} = \text{local}(\mathbf{T})$
+2. **Retract or retraction**: The opposite operation: mapping from the tangent space back to the manifold $\mathbf{T} = \text{retract}(\mathbf{\xi})$.
```
Likewise, the more general retract operations needs the input of the manifold object from which the tangent space is referenced from.

It looks like you are prematurely trying to make manifolds looks like lie groups.  Where as you could instead explain it as  we  have objects with differentiable properties (manifolds) and objects with composition/inverse/identity properties (groups), and when they have both we get Lie groups.  (where the retract and local operations are conveniently defined with respect to the group Identity).

Also please stress that groups are not needed for most of GTSAM.     Its just that when the manifold object are also groups, we can express the local and retract operations with respect to the identity object.

mmattamala replied:
Thanks for pointing this out. I hope this new version more explicit and consistent with GTSAM.
Now Lie groups should mainly be important in the part 3 only, when we use the adjoint to transform covariances.

mmattamala commented on _posts/2021-02-07-uncertainties-part2.md:
```diff
@@ -0,0 +1,631 @@
+\end{equation}
+$$
+
+so we can apply the transformation on the left-hand side:
```
I added an explanation in part 3, when we talk about covariance transformations, to explicitly mention this case and how it's related to the adjoint

mmattamala commented on _posts/2021-02-07-uncertainties-part2.md:
```diff
@@ -0,0 +1,631 @@
+
+(Please note here that we used *capitalized* $$\text{Log}(\cdot) := \text{log}( \cdot)^{\vee}$$ and $$\text{Exp}(\cdot):=\text{exp}( (\cdot)^{\wedge})$$ operators for simplicity as used by   [Forster et al (2017),](https://arxiv.org/abs/1512.02363) and [Solà et al. (2020)](https://arxiv.org/abs/1812.01537).)
+
+The main difference with Lie groups is that by using the general concept of [_retraction_](https://press.princeton.edu/absil) we can use **alternative definitions to go from the tangent space to the manifold, that can be more efficient than the exponential map when solving optimization problems.**
```
I added a paragraph to explain this as well.

mmattamala commented on _posts/2021-02-07-uncertainties-part2.md:
```diff
@@ -0,0 +1,631 @@
+    alt="Hessian matrix" />
+    <figcaption>With left-hand convention the increments are applied from the left, changing our reference frame to the new robot pose $B_{i+1}$ (in black) .</figcaption>
+</figure>
+<br />
```
Solved in part 3 I hope

dellaert commented:
Ready to go?

mmattamala commented:
Yes, thanks Frank!

michaelbosse commented:
Could someone post the updated pdfs again so I can quickly see the changes?

mmattamala commented:
Hi Mike, here you go. Thanks for your help!

[Reducing the uncertainty about the uncertainties, part 1_ Linear and nonlinear _ GTSAM.pdf](https://github.com/borglab/gtsam.org/files/6030673/Reducing.the.uncertainty.about.the.uncertainties.part.1_.Linear.and.nonlinear._.GTSAM.pdf)

[Reducing the uncertainty about the uncertainties, part 2_ Frames and manifolds _ GTSAM.pdf](https://github.com/borglab/gtsam.org/files/6030674/Reducing.the.uncertainty.about.the.uncertainties.part.2_.Frames.and.manifolds._.GTSAM.pdf)

[Reducing the uncertainty about the uncertainties, part 3_ Adjoints and covariances _ GTSAM.pdf](https://github.com/borglab/gtsam.org/files/6030677/Reducing.the.uncertainty.about.the.uncertainties.part.3_.Adjoints.and.covariances._.GTSAM.pdf)

dellaert commented:
I'll await Mike's OK and then merge after I make sure it renders well. Then I'll tweet massively about it, and you should re-tweet and get others to re-tweet :-) Alternatively, *you* can tweet and I'll ret-tweet. I would recommend making a small tweet-storm with the skinny, as not many people will actually take the time to go to the blog itself without that "teaser" material.

michaelbosse commented:
Overall this reads much better!

Just some minor nits:
- Eqns 29+30, Part II, and Eqns 17+24, Part III are too long and get clipped by the page width.
- I'm not sure we need to specify the same frame as a prefix and suffix for tangent space vectors; seems redundant since they are both always the same.  IMHO the extra symbols unnecessarily clutter the equations making it harder to see the elegance in the math.
- The latest release of GTSAM doesn't default to S0(3)xR^3 anymore but rather showed that SE(3) converges better in optimizations for minimal (if any) performance loss.

mmattamala commented:
Awesome, thanks a lot for going through the whole thing! I'll make the last changes tomorrow morning (in UK time):

- I'll see if I can change the equations a bit to fit the space; I know that in the website there should appear a scroll bar to get access to the whole expression, but I'd like to know if @dellaert has a better suggestion.
- Yeah, I agree that makes the expressions more cluttered and thanks for pointing that out. I'll stay with the prefix for the vectors to be consistent with previous parts (like the camera example)
- I think it's better if I just remove that part in which I mention SO(3) X R^3 to not make things more complicated if GTSAM simply uses the exponential map for that.

After that I guess we can agree the post should be ready for Frank to check how it renders and we start the tweet-storm :)

mmattamala commented:
@dellaert I ended up fixing it now, so feel free to check how it renders and proceed :)

jlblancoc commented:
1 sec! I was just about to post my two cents... ;-) 

[1st part](https://drive.google.com/file/d/1S6BG-wn18twatZ4ONRZLDXAGzzBSTaWW/view?usp=sharing)

(more to come)

mmattamala commented:
@jlblancoc haha great! we can postpone this a bit to give you more time and check your comments carefully, thanks a lot.

jlblancoc commented:
[2nd part](https://drive.google.com/file/d/1SHrumZ_4U27eD7VIQ6SUUZ1ARJOF-IPV/view?usp=sharing)

@mmattamala  thank you for undertaking the ambitious task of explaining this material! Legions of future students will thank you ;-)

jlblancoc commented:
And the last one: [3rd part](https://drive.google.com/file/d/1SJOrY_0x-2zpduf-VHLPnyRgTWi5dq8Q/view?usp=sharing)

Congrats, it's an excellent material, I really believe in the need to visually express all these concepts with pictures... a *must read* for GTSAM users, and PhD students in our field in general.

mmattamala commented:
Thanks a lot @jlblancoc ! I really appreciate your words :D
I think I addressed all your comments, I've attached the PDFs.

@dellaert I think it's ready, feel free to do the final check.

[Reducing the uncertainty about the uncertainties, part 1_ Linear and nonlinear _ GTSAM.pdf](https://github.com/borglab/gtsam.org/files/6038686/Reducing.the.uncertainty.about.the.uncertainties.part.1_.Linear.and.nonlinear._.GTSAM.pdf)

[Reducing the uncertainty about the uncertainties, part 2_ Frames and manifolds _ GTSAM.pdf](https://github.com/borglab/gtsam.org/files/6038710/Reducing.the.uncertainty.about.the.uncertainties.part.2_.Frames.and.manifolds._.GTSAM.pdf)

[Reducing the uncertainty about the uncertainties, part 3_ Adjoints and covariances _ GTSAM.pdf](https://github.com/borglab/gtsam.org/files/6038688/Reducing.the.uncertainty.about.the.uncertainties.part.3_.Adjoints.and.covariances._.GTSAM.pdf)

dellaert commented:
I checked it on my mac and it renders well, so i'll merge. Amazing series of posts!
